### PR TITLE
feat(azure): implemented azure branch status checks for automerge

### LINKED ISSUE #7392: [azure] Fully implement branch statuses
Azure devops provides a git commit status [api](https://docs.microsoft.com/en-us/rest/api/azure/devops/git/statuses?view=azure-devops-rest-4.1), so we should implement this.

Without this feature it isn't possible for renovate to wait for succeding pipelines or setting custom statuses.

ref #1149

/cc @zleight1

### PR DESCRIPTION
**Pull request by zleight1**
Implemented checks for branch statuses for azure, using the latest branch gitStatusState to determine if it's pending, failing or successful. This is required to properly automerge, as it was always pending prior.
Added tests for scenarios with various results and updated relevant existing tests.

Closes #7392

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
Please see other platforms for implementation. You need to take required status check config option into account.

**Comment by zleight1** [comment]
> Please see other platforms for implementation. You need to take required status check config option into account.

Okay, I wasn't sure because the docs on renovate say it's not supported so I didn't implement it in this PR.

**Comment by zleight1** [comment]
> Yes, you don't need to fully implement that. Please check the other platforms for the checks.

Not sure I understand when you say check the other platforms for checks, you mean the unit tests or something else?

**Comment by viceice** [comment]
Eg, see here

https://github.com/renovatebot/renovate/blob/765786a81a3f70f5e224ebb2a885a33fb9c93523/lib/platform/bitbucket-server/index.ts#L432

**Comment by zleight1** [comment]
@viceice I tried to mirror the other implementations as closely as possible, I also noticed where that status check with context is used and it needs a setter to match so I implemented that as well.

File: lib/platform/azure/index.ts [hunk 1]
```diff
@@ -504,17 +553,39 @@ export async function ensureCommentRemoval({
...
+    state: renovateToAzureStatusMapping[state],
+    targetUrl,
+  };
+  await azureApiGit.createCommitStatus(
+    statusToCreate,
+    branch.commit.commitId,
+    config.repoId
+  );
   logger.debug(
     `setBranchStatus(${branchName}, ${context}, ${description}, ${state}, ${targetUrl}) - Not supported by Azure DevOps (yet!)`
```
Comment (JamieMagee) [inline lib/platform/azure/index.ts:587]:
```suggestion
    `setBranchStatus(${branchName}, ${context}, ${description}, ${state}, ${targetUrl})`
```

Not true anymore! :smile:

File: lib/platform/azure/index.ts [hunk 2]
```diff
@@ -282,20 +286,44 @@ export async function getBranchPr(branchName: string): Promise<Pr | null> {
...
+  // only grab the latest statuses, it will group any by context
+  const statuses = await azureApiGit.getStatuses(
+    branch.commit.commitId,
+    config.repoId,
+    undefined,
+    undefined,
+    undefined,
+    true
+  );
+  return statuses;
```
Comment (JamieMagee) [inline lib/platform/azure/index.ts:304]:
```suggestion
  return await azureApiGit.getStatuses(
    branch.commit.commitId,
    config.repoId,
    undefined,
    undefined,
    undefined,
    true
  );
```

**Comment by JamieMagee** [comment]
@zleight1 Have you gotten a chance to test this against Azure DevOps? If not, I can do it tomorrow.

P.S. apologies about closing the PR. I hit the wrong button!

File: lib/platform/azure/index.ts (hunk 1 above)
Comment (zleight1) [reply lib/platform/azure/index.ts]: Nice catch, I also moved it up to the beginning of the function

**Comment by zleight1** [comment]
> @zleight1 Have you gotten a chance to test this against Azure DevOps? If not, I can do it tomorrow.
> 
> P.S. apologies about closing the PR. I hit the wrong button!

I'll have time to check it on Sunday hopefully, if you could check it earlier that would be great.

**Comment by JamieMagee** [comment]
So I managed to test this, and the status checks are indeed working. They're not surfaced in PRs, but I'm not sure if that's expected—there is a separate pull request status API as well.

image

File: lib/platform/azure/index.ts [hunk 3]
```diff
@@ -267,20 +271,43 @@ export async function getBranchPr(branchName: string): Promise<Pr | null> {
...
+async function getStatusCheck(branchName: string): Promise<GitStatus[]> {
   const azureApiGit = await azureApi.gitApi();
   const branch = await azureApiGit.getBranch(
     config.repoId,
     getBranchNameWithoutRefsheadsPrefix(branchName)
   );
-  if (branch.aheadCount === 0) {
-    return BranchStatus.green;
+  // only grab the latest statuses, it will group any by context
+  return await azureApiGit.getStatuses(
```
Comment (JamieMagee) [inline lib/platform/azure/index.ts:281]: I made this suggestion earlier, and it broke an eslint rule :facepalm: 

```suggestion
  return azureApiGit.getStatuses(
```

**Review by JamieMagee — APPROVED** [review]

**Pull request merged**